ehrql/query_engines/base_sql.py

```diff
@@ -728,7 +773,16 @@ def reify_query(self, query):
         e.g. using `.alias()` to make a sub-query, using `.cte()` to make a Common Table
         Expression, or writing the results of the query to a temporary table.
         """
-        return query.cte(name=f"cte_{self.get_next_id()}")
+        cte = query.cte(name=f"cte_{self.get_next_id()}")
```

Comment:
The trino query_engine has its own `reify_query` method that overrides this one - does it need to handle the no_replacement_traverse as well?

Reply:
The problem is specifically with CTEs, not other kinds of table-like thing which SQLAlchemy generally considers immutable.

I did consider refactoring things so that `reify_query()` always applies this flag and then calls out to some other `actually_reify_query()` function which subclasses can override. But my thinking was that the test case will catch any future instances of this problem and then we can apply the flag as it's needed and avoid complicating the call stack here.